vm/src/deserializer.h

```diff
@@ -0,0 +1,83 @@
+#ifndef BORGES_DESERIALIZER_H
+#define BORGES_DESERIALIZER_H
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace Borges {
+
+class Deserializer {
+public:
+    Deserializer(std::size_t n, const double* p);
+    void deserialize(double* p);
+    template <typename T>
+    void deserialize(std::shared_ptr<T>* p);
+
+private:
+    std::size_t size_;
+    const double* data_;
+    std::size_t index_;
+    std::unordered_map<std::size_t, std::shared_ptr<void>> index_to_ptr_map_;
+};
+
+void deserialize(Deserializer& d, bool* p);
+void deserialize(Deserializer& d, double* p);
+void deserialize(Deserializer& d, std::size_t* p);
+
+template <typename T>
+T
+deserialize(Deserializer& d)
+{
+    alignas(T) char b[sizeof(T)];
+    T* p = reinterpret_cast<T*>(b);
```

Comment:
I would write this as reinterpret_cast<T*>(&b[0]) for clarity

Reply:
That seems to suggest that p is a pointer to the first element in the buffer, whereas it should more properly be thought of as a pointer to the entire buffer. Would this work for you as an alternative?
reinterpret_cast<T*>(&b)

Reply:
no, because b is already a char *. you actually *do* want a pointer to the first element. this way, you'd take the address of the address of b.

Reply:
You are right, but in that case I disagree that reinterpret_cast<T*>(&b[0]) adds clarity. In fact, it does the opposite for me: you correctly pointed out that b is already a char*, and this simply a pointer cast, but its precisely the change you proposed that got me confused about this fact.

That's placement new syntax. It works exactly like the new operator you are used to, except that it doesn't allocate memory for the object to be constructed. Instead, it uses the memory pointed to by p to construct the shared_ptr in

Reply:
to me, it *is* more clear because you make explicit that yo take the address of the beginning of the array. just passing b is basically an implicit cast to * of the value of b on the stack.